src/Sarif.UnitTests/Visitors/ValidatingVisitorTests.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System.IO;
+using FluentAssertions;
+using System.Text;
+using Newtonsoft.Json;
+using Microsoft.CodeAnalysis.Sarif.Readers;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+    [TestClass]
+    public class ValidatingVisitorTests
+    {
+        [TestMethod]
+        public void ValidatingVisitor_InvalidUri()
+        {
+            // NOTE that for validating tests, we need to deserialize SARIF text 
+            // without benefit of the many converters that transform specified
+            // JSON values to enum equivalents, etc. The Sarif text below and 
+            // for all these tests is rendered, in some cases, as it the converter
+            // had done some processing before conversion to eventual type. We
+            // express the version as OneZeroZeroBeta5, for example, because the
+            // default JSON.NET enum conversion will do the right thing as far
+            // as converting this value to the C# OM.
+            string input =
+@"{
+  ""version"": ""OneZeroZeroBetaFive"",
+  ""runs"": [{
+      ""tool"": { ""name"": ""toolName"" },
+      ""rules"" : { ""id1"" : { ""id"" : ""id1"", ""helpUri"" : ""??[][][]??""} },
+      ""files"": { ""!!not a valid uri!!"" : { ""uri"" : ""!!not a valid uri!!""} },
+      ""results"": [{
+            ""locations"": [{""resultFile"": { ""uri"": ""file:///C:/space space.cpp"" }}],
+            ""stacks"": [{ ""frames"" : [{ ""uri"" : ""c:\\test.cpp"", ""fullyQualifiedLogicalName"" : ""fqn"" }]}],
+            ""fixes"": [{ ""fileChanges"": [{ ""uri"" : ""file://c:/missing/slash"", ""replacements"" :[{ ""offset"" : 0, ""deletedLength"" : 1 }] }], ""description"" : ""ok""}]
+        }]
+      }]
+}";
+
+            // TODO do we want to validate that http:// Uris have a trailing slash?
```

Comment:
My point being, we can't know that a particular URI we find in a file would behave the same if a slash were added to it, so our validator can't insist on it.

---
In reply to: [64092768](https://github.com/Microsoft/sarif-sdk/pull/296#discussion_r64092768) [](ancestors = 64092768)